### PR TITLE
Don't zero non-readable buffer regions when capacity is decreased

### PR DESCRIPTION
**Pull request by njhill**
Motivation

#1802 fixed `ByteBuf` implementations to ensure that the whole buffer region is preserved when capacity is increased, not just the readable part. The behaviour is still different however when the capacity is _decreased_ - data outside the currently-readable region is zeroed.

Modifications

Update `ByteBuf#capacity(int)` implementations to also copy the whole buffer region when the new capacity is less than the current capacity.

Result

Consistent behaviour of `ByteBuf#capacity(int)` regardless of whether the new capacity is greater than or less than the current capacity.

### CODE REVIEW

**Comment by njhill** [comment]
Added a commit to:
- Also cover the case that the newCapacity <= readerIndex
- Simplify and normalize the three implementations (less code)
- Eliminate some redundant checks/index updates in `PoolArena#reallocate(...)`

File: buffer/src/main/java/io/netty/buffer/UnpooledHeapByteBuf.java [hunk 1]
```diff
@@ -120,29 +120,28 @@ public int capacity() {
...
+            if (readerIndex() > newCapacity) {
                 setIndex(newCapacity, newCapacity);
+            } else if (writerIndex() > newCapacity) {
+                writerIndex(newCapacity);
             }
-            setArray(newArray);
-            freeArray(oldArray);
         }
+        byte[] newArray = allocateArray(newCapacity);
+        //TODO Arrays.copyOf would be preferable here
```
Comment (normanmaurer) [inline buffer/src/main/java/io/netty/buffer/UnpooledHeapByteBuf.java]: Then why not use it ?

**Comment by normanmaurer** [comment]
@netty-bot test this please

File: buffer/src/main/java/io/netty/buffer/UnpooledHeapByteBuf.java (hunk 1 above)
Comment (franz1981) [reply buffer/src/main/java/io/netty/buffer/UnpooledHeapByteBuf.java]: Was going to write the same...

Comment (njhill) [reply buffer/src/main/java/io/netty/buffer/UnpooledHeapByteBuf.java]: Wouldn't it break subclasses that rely on the use of `allocateArray()` e.g. to [track buffer heap usage](https://github.com/netty/netty/blob/4.1/buffer/src/main/java/io/netty/buffer/UnpooledByteBufAllocator.java#L157) or impl [custom pooling](https://github.com/netty/netty/pull/8015)?

Though I noticed that the `copy()` method currently violates this, I can open another PR to fix that.

**Comment by njhill** [comment]
Added a commit with further small simplification/de-duplication.

File: buffer/src/main/java/io/netty/buffer/UnpooledHeapByteBuf.java (hunk 1 above)
Comment (njhill) [reply buffer/src/main/java/io/netty/buffer/UnpooledHeapByteBuf.java]: See #9440 for the other mentioned fix

Comment (normanmaurer) [reply buffer/src/main/java/io/netty/buffer/UnpooledHeapByteBuf.java]: I see.... Makes sense. Can you remove TODO: then ?

Comment (njhill) [reply buffer/src/main/java/io/netty/buffer/UnpooledHeapByteBuf.java]: @normanmaurer done

**Comment by normanmaurer** [comment]
@netty-bot test this please

**Comment by normanmaurer** [comment]
@njhill thanks a lot!

**Pull request merged**